### PR TITLE
Android: support exit procedure

### PR DESCRIPTION
**Pull request by bofeng-song**

### CODE REVIEW

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 1]
```diff
@@ -27,20 +27,20 @@
 
 #include "base/Log.h"
 #include "base/memory/Memory.h"
-#include "platform/android/AndroidPlatform.h"
+#include "cocos/bindings/event/CustomEventTypes.h"
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: cocos/bindings/event/CustomEventTypes.h -> bindings/event/CustomEventTypes.h

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 2]
```diff
@@ -366,13 +379,18 @@ class GameInputProxy {
...
+    }
+
 private:
     AndroidPlatform *_androidPlatform{nullptr};
     JNIEnv *_jniEnv{nullptr};         // JNI environment
     int32_t _gameControllerIndex{-1}; // Most recently connected game controller index
     bool _running{false};
     bool _isVisible{false};
     bool _hasWindow{false};
+    bool _recheckDestroy{true};
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: what does it mean

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 3]
```diff
@@ -414,14 +432,20 @@ AndroidPlatform::~AndroidPlatform() {
...
+    }
     _inputProxy = new GameInputProxy(this);
-    registerInterface(std::make_shared<Accelerometer>());
-    registerInterface(std::make_shared<Battery>());
-    registerInterface(std::make_shared<Network>());
-    registerInterface(std::make_shared<Screen>());
-    registerInterface(std::make_shared<System>());
-    registerInterface(std::make_shared<SystemWindow>());
-    registerInterface(std::make_shared<Vibrator>());
+    _inputProxy->setRunning(isRunning);
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: isRunning is not a good name, may be should use isInputValid

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 4]
```diff
@@ -414,14 +432,20 @@ AndroidPlatform::~AndroidPlatform() {
 }
 
 int AndroidPlatform::init() {
+    bool isRunning = _inputProxy != nullptr;
+    if (!_inputProxy) {
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: !_inputProxy -> ! isInitialized()

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 5]
```diff
@@ -61,9 +62,24 @@
     }
 
 #define INPUT_ACTION_COUNT 6
+#define COCOS_EXIT_GAME 101
 
 extern int cocos_main(int argc, const char **argv); // NOLINT(readability-identifier-naming)
 
+extern "C" {
+bool perform_in_game_thread(struct android_app *app, cc::AndroidPlatform::GameThreadFunc func) {
```
Comment (dumganhar) [inline native/cocos/platform/android/AndroidPlatform.cpp]: Where is `perform_in_game_thread` invoked?

Comment (bofeng-song) [reply native/cocos/platform/android/AndroidPlatform.cpp]: invoke from here:  https://github.com/cocos/cocos-engine-external/pull/253

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 6]
```diff
@@ -512,8 +512,8 @@ ScriptEngine::ScriptEngine()
 ScriptEngine::~ScriptEngine() {
     #if !CC_EDITOR
     if (gSharedV8) {
-        delete gSharedV8;
-        gSharedV8 = nullptr;
+//        delete gSharedV8;
```
Comment (minggo) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: Please add a comment why not delete it.

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 7]
```diff
@@ -431,9 +470,20 @@ int AndroidPlatform::getSdkVersion() const {
...
+    cc::FileUtilsAndroid::setassetmanager(app->activity->assetManager);
+    _app = app;
+    CC_SAFE_DELETE(_inputProxy);
+    _inputProxy = ccnew GameInputProxy(this);
+    _inputProxy->setLaunched(_isInited);
+    _app->userData = _inputProxy;
+    _app->onAppCmd = handleCmdProxy;
+
+    if (!_isInited) {
+        init();
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: It is better to check `_isInited` in `init()`.

File: native/cocos/bindings/manual/jsb_global_init.cpp [hunk 8]
```diff
@@ -197,6 +197,8 @@ void jsb_init_file_operation_delegate() { //NOLINT
         CC_ASSERT(delegate.isValid());
 
         se::ScriptEngine::getInstance()->setFileOperationDelegate(delegate);
+    } else {
+        se::ScriptEngine::getInstance()->setFileOperationDelegate(delegate); // game maybe relaunched at different thread but at the same process
```
Comment (minggo) [inline native/cocos/bindings/manual/jsb_global_init.cpp]: When does this happen?

File: native/cocos/platform/android/AndroidPlatform.h [hunk 9]
```diff
@@ -58,9 +58,9 @@ class CC_DLL AndroidPlatform : public UniversalPlatform {
 
     int32_t getHeight() const;
 
-    void setAndroidApp(android_app *app);
-
 private:
+    bool _isInited{false};
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.h]: Will it be initialized again?

File: native/cocos/platform/android/jni/JniCocosEntry.cpp [hunk 10]
```diff
@@ -26,20 +26,15 @@
...
 
 void android_main(struct android_app *app) {
-    cc::FileUtilsAndroid::setassetmanager(app->activity->assetManager);
-
     auto *platform = cc::BasePlatform::getPlatform();
     auto *androidPlatform = static_cast<cc::AndroidPlatform *>(platform);
-    androidPlatform->setAndroidApp(app);
-    platform->init();
-    platform->loop();
+    androidPlatform->run(0, reinterpret_cast<const char**>(app));
```
Comment (minggo) [inline native/cocos/platform/android/jni/JniCocosEntry.cpp]: Should pass 1, as it has one parameter.

File: native/cocos/platform/android/AndroidPlatform.h (hunk 9 above)
Comment (bofeng-song) [reply native/cocos/platform/android/AndroidPlatform.h]: Yes, game launched at the same process will trigger it.

File: native/cocos/bindings/manual/jsb_global_init.cpp (hunk 8 above)
Comment (bofeng-song) [reply native/cocos/bindings/manual/jsb_global_init.cpp]: While game launch from recent task, but gameActivity has been destroy by user or system

Comment (minggo) [reply native/cocos/bindings/manual/jsb_global_init.cpp]: Got it, please add it in comment.

File: native/cocos/platform/android/jni/JniCocosEntry.cpp [hunk 11]
```diff
@@ -26,20 +26,15 @@
...
 
 void android_main(struct android_app *app) {
-    cc::FileUtilsAndroid::setassetmanager(app->activity->assetManager);
-
     auto *platform = cc::BasePlatform::getPlatform();
     auto *androidPlatform = static_cast<cc::AndroidPlatform *>(platform);
-    androidPlatform->setAndroidApp(app);
-    platform->init();
-    platform->loop();
+    androidPlatform->run(1, reinterpret_cast<const char**>(app));
```
Comment (minggo) [inline native/cocos/platform/android/jni/JniCocosEntry.cpp]: It is more clear to do like this
```c++
    androidPlatform->init();
    androidPlatform->run(1, reinterpret_cast<const char**>(app));
```

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 12]
```diff
@@ -430,10 +454,19 @@ int AndroidPlatform::getSdkVersion() const {
...
-    }
+int32_t AndroidPlatform::run(int  /*argc*/, const char **argv) {
+    auto *app = reinterpret_cast<struct android_app *>(argv);
+    cc::FileUtilsAndroid::setassetmanager(app->activity->assetManager);
+    _app = app;
+    CC_SAFE_DELETE(_inputProxy);
+    _inputProxy = ccnew GameInputProxy(this);
+    _inputProxy->setLaunched(_isInited);
+    _app->userData = _inputProxy;
+    _app->onAppCmd = handleCmdProxy;
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: It is better to put these logic into `init()`
```c++
auto *app = reinterpret_cast<struct android_app *>(argv);
    cc::FileUtilsAndroid::setassetmanager(app->activity->assetManager);
    _app = app;
    _inputProxy = ccnew GameInputProxy(this);
    _inputProxy->setLaunched(_isInited);
    _app->userData = _inputProxy;
    _app->onAppCmd = handleCmdProxy;
```

Comment (minggo) [reply native/cocos/platform/android/AndroidPlatform.cpp]: Then add a `destroy()`, and delete `_inputProxy` and unregister interfaces in it.

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 13]
```diff
@@ -430,10 +454,19 @@ int AndroidPlatform::getSdkVersion() const {
...
+
+    init();
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: Do initialization in run is not good.

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 14]
```diff
@@ -509,14 +509,8 @@ ScriptEngine::ScriptEngine()
     #endif
 }
 
-ScriptEngine::~ScriptEngine() {
-    #if !CC_EDITOR
-    if (gSharedV8) {
-        delete gSharedV8;
-        gSharedV8 = nullptr;
```
Comment (minggo) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp:504]: Should only don't delete it on Android. And please add some comment about the reason.

File: native/cocos/platform/BasePlatform.h [hunk 15]
```diff
@@ -160,6 +160,10 @@ class CC_DLL BasePlatform {
         _osInterfaces.erase(it);
     }
 
+    void unregisterAllInterface() {
```
Comment (dumganhar) [inline native/cocos/platform/BasePlatform.h]: unregisterAllInterface -> unregisterAllInterfaces

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 16]
```diff
@@ -312,9 +312,8 @@ class GameInputProxy {
                 WindowEvent ev;
                 ev.type = WindowEvent::Type::CLOSE;
                 _androidPlatform->dispatchEvent(ev);
-                _androidPlatform->_isInited = false;
                 _androidPlatform->onDestory();
-                CC_SAFE_DELETE(_androidPlatform->_inputProxy);
+                _androidPlatform->destroy();
```
Comment (dumganhar) [inline native/cocos/platform/android/AndroidPlatform.cpp]: _androidPlatform->onDestory();
_androidPlatform->destroy();

??

Comment (bofeng-song) [reply native/cocos/platform/android/AndroidPlatform.cpp]: onDestroy will trigger Application's destructor
destroy can be consider to put to engine's destructor

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 17]
```diff
@@ -509,14 +509,22 @@ ScriptEngine::ScriptEngine()
     #endif
 }
 
+#if CC_PLATFORM == CC_PLATFORM_ANDROID
+/**
+ * v8::V8::Initialize() can only be called once time at same process.
```
Comment (minggo) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: be called once time -> be called once

File: native/cocos/platform/android/AndroidPlatform.cpp (hunk 16 above)
Comment (minggo) [reply native/cocos/platform/android/AndroidPlatform.cpp]: Yep, the logic is weird.

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 18]
```diff
@@ -497,14 +497,22 @@ ScriptEngine::ScriptEngine()
     #endif
 }
 
+#if CC_PLATFORM == CC_PLATFORM_ANDROID
+/**
+ * v8::V8::Initialize() can only be called once for a process.
+ * After calling onDestroy on Android platform, the process will be maintained for a period of time. so we disable gSharedV8 to release here.
```
Comment (dumganhar) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: > disable gSharedV8 to release here.

??

**Review by minggo — APPROVED** [review]

**Review by dumganhar — APPROVED** [review]

**Pull request merged**